File: code/backend/src/Services/SmtpMailer.php
<?php

declare(strict_types=1);

namespace BudgetCentre\Services;

use BudgetCentre\Support\Env;
use RuntimeException;

final class SmtpMailer
{
    public function send(string $to, string $subject, string $body): void
    {
        $host = Env::string('SMTP_HOST');
        $port = Env::int('SMTP_PORT', 465);
        $username = Env::string('SMTP_USERNAME');
        $password = Env::string('SMTP_PASSWORD');
        $from = Env::string('MAIL_FROM', $username);
        $fromName = Env::string('MAIL_FROM_NAME', 'BudgetCentre');

        if ($host === null || $username === null || $password === null || $from === null) {
            throw new RuntimeException('SMTP is not configured.');
        }

        $socket = stream_socket_client(
            "ssl://{$host}:{$port}",
            $errno,
            $error,
            15,
        );

        if ($socket === false) {
            throw new RuntimeException("SMTP connection failed: {$error}", $errno);
        }

        try {
            $this->expect($socket, 220);
            $this->command($socket, 'EHLO budgetcentre.local', 250);
            $this->command($socket, 'AUTH LOGIN', 334);
            $this->command($socket, base64_encode($username), 334);
            $this->command($socket, base64_encode($password), 235);
            $this->command($socket, "MAIL FROM:<{$from}>", 250);
            $this->command($socket, "RCPT TO:<{$to}>", [250, 251]);
            $this->command($socket, 'DATA', 354);

            $headers = [
                'From: ' . $this->formatAddress($from, $fromName),
                "To: <{$to}>",
                'Subject: ' . $this->encodeHeader($subject),
                'MIME-Version: 1.0',
                'Content-Type: text/plain; charset=UTF-8',
                'Content-Transfer-Encoding: 8bit',
            ];
            fwrite($socket, implode("\r\n", $headers) . "\r\n\r\n" . $this->escapeBody($body) . "\r\n.");
            fwrite($socket, "\r\n");
            $this->expect($socket, 250);
            $this->command($socket, 'QUIT', 221);
        } finally {
            fclose($socket);
        }
    }

    /**
     * @param resource $socket
     * @param int|array<int> $expected
     */
    private function command(mixed $socket, string $command, int|array $expected): void
    {
        fwrite($socket, $command . "\r\n");
        $this->expect($socket, $expected);
    }

    /**
     * @param resource $socket
     * @param int|array<int> $expected
     */
    private function expect(mixed $socket, int|array $expected): void
    {
        $expectedCodes = is_array($expected) ? $expected : [$expected];
        $response = '';

        do {
            $line = fgets($socket);
            if ($line === false) {
                throw new RuntimeException('SMTP server closed the connection.');
            }
            $response .= $line;
        } while (isset($line[3]) && $line[3] === '-');

        $code = (int) substr($response, 0, 3);
        if (!in_array($code, $expectedCodes, true)) {
            throw new RuntimeException("SMTP command failed: {$response}");
        }
    }

    private function formatAddress(string $email, string $name): string
    {
        return $this->encodeHeader($name) . " <{$email}>";
    }

    private function encodeHeader(string $value): string
    {
        return '=?UTF-8?B?' . base64_encode($value) . '?=';
    }

    private function escapeBody(string $body): string
    {
        return preg_replace('/^\./m', '..', $body) ?? $body;
    }
}

File: code/backend/internal/app/smtp_mailer.go
package app

import (
	"bufio"
	"crypto/tls"
	"encoding/base64"
	"fmt"
	"net"
	"net/mail"
	"net/smtp"
	"strconv"
	"strings"
	"time"
)

func (a *App) sendMail(to, subject, body string) error {
	if a.cfg.SMTPHost == "" || a.cfg.SMTPUsername == "" || a.cfg.SMTPPassword == "" {
		return fmt.Errorf("smtp is not configured")
	}
	from := a.cfg.MailFrom
	if from == "" {
		from = a.cfg.SMTPUsername
	}
	safeFrom, err := sanitizeSMTPAddress(from)
	if err != nil {
		return err
	}
	safeTo, err := sanitizeSMTPAddress(to)
	if err != nil {
		return err
	}
	safeSubject := sanitizeSMTPHeaderValue(subject)
	if a.cfg.SMTPPort == 465 {
		return a.sendMailTLS(safeTo, safeFrom, safeSubject, body)
	}
	auth := smtp.PlainAuth("", a.cfg.SMTPUsername, a.cfg.SMTPPassword, a.cfg.SMTPHost)
	return smtp.SendMail(a.cfg.SMTPHost+":"+strconv.Itoa(a.cfg.SMTPPort), auth, safeFrom, []string{safeTo}, []byte(mailMessage(safeFrom, a.cfg.MailFromName, safeTo, safeSubject, body)))
}

func (a *App) sendMailTLS(to, from, subject, body string) error {
	address := a.cfg.SMTPHost + ":" + strconv.Itoa(a.cfg.SMTPPort)
	dialer := net.Dialer{Timeout: 15 * time.Second}
	conn, err := tls.DialWithDialer(&dialer, "tcp", address, &tls.Config{ServerName: a.cfg.SMTPHost})
	if err != nil {
		return err
	}
	defer conn.Close()
	reader := bufio.NewReader(conn)
	if err := smtpExpect(reader, 220); err != nil {
		return err
	}
	commands := []struct {
		command  string
		expected []int
	}{
		{"EHLO budgetcentre.local", []int{250}},
		{"AUTH LOGIN", []int{334}},
		{base64.StdEncoding.EncodeToString([]byte(a.cfg.SMTPUsername)), []int{334}},
		{base64.StdEncoding.EncodeToString([]byte(a.cfg.SMTPPassword)), []int{235}},
		{"MAIL FROM:<" + from + ">", []int{250}},
		{"RCPT TO:<" + to + ">", []int{250, 251}},
		{"DATA", []int{354}},
	}
	for _, item := range commands {
		if _, err := fmt.Fprintf(conn, "%s\r\n", item.command); err != nil {
			return err
		}
		if err := smtpExpect(reader, item.expected...); err != nil {
			return err
		}
	}
	if _, err := fmt.Fprintf(conn, "%s\r\n.\r\n", escapeSMTPBody(mailMessage(from, a.cfg.MailFromName, to, subject, body))); err != nil {
		return err
	}
	if err := smtpExpect(reader, 250); err != nil {
		return err
	}
	_, _ = fmt.Fprint(conn, "QUIT\r\n")
	_ = smtpExpect(reader, 221)
	return nil
}

func smtpExpect(reader *bufio.Reader, expected ...int) error {
	var response string
	for {
		line, err := reader.ReadString('\n')
		if err != nil {
			return err
		}
		response += line
		if len(line) < 4 || line[3] != '-' {
			break
		}
	}
	code, _ := strconv.Atoi(strings.TrimSpace(response[:3]))
	for _, item := range expected {
		if code == item {
			return nil
		}
	}
	return fmt.Errorf("smtp command failed: %s", response)
}

func mailMessage(from, fromName, to, subject, body string) string {
	headers := []string{
		"From: " + encodedHeader(fromName) + " <" + from + ">",
		"To: <" + to + ">",
		"Subject: " + encodedHeader(subject),
		"MIME-Version: 1.0",
		"Content-Type: text/plain; charset=UTF-8",
		"Content-Transfer-Encoding: 8bit",
	}
	return strings.Join(headers, "\r\n") + "\r\n\r\n" + body
}

func sanitizeSMTPAddress(value string) (string, error) {
	if strings.ContainsAny(value, "\r\n") {
		return "", fmt.Errorf("invalid email address")
	}
	parsed, err := mail.ParseAddress(strings.TrimSpace(value))
	if err != nil || parsed.Address == "" || strings.ContainsAny(parsed.Address, "\r\n") {
		return "", fmt.Errorf("invalid email address")
	}
	return parsed.Address, nil
}

func sanitizeSMTPHeaderValue(value string) string {
	value = strings.ReplaceAll(value, "\r", "")
	value = strings.ReplaceAll(value, "\n", "")
	return strings.TrimSpace(value)
}

func encodedHeader(value string) string {
	if value == "" {
		value = "BudgetCentre"
	}
	return "=?UTF-8?B?" + base64.StdEncoding.EncodeToString([]byte(value)) + "?="
}

func escapeSMTPBody(body string) string {
	lines := strings.Split(body, "\n")
	for i, line := range lines {
		if strings.HasPrefix(line, ".") {
			lines[i] = "." + line
		}
	}
	return strings.Join(lines, "\n")
}
